### PR TITLE
Feature/pythonscriptsource

### PR DESCRIPTION
**Pull request by martinfalk**
exposed DataMapper to inviwopy, extended pyvolume interface, and added PythonScriptSource processor

### CODE REVIEW

File: modules/python3/processors/pythonscriptsource.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+void PythonScriptSource::process() {
+    if (script_.getSource().empty()) {
+        mesh_.setData(std::make_shared<Mesh>());
+        return;
+    }
+    script_.run({}, [&](pybind11::dict dict) {
+        if (dict.contains("mesh")) {
+            auto pyMesh = dict["mesh"];
+            auto mesh = std::shared_ptr<BasicMesh>(pyMesh.cast<BasicMesh*>());
```
Comment (ResearchDaniel) [inline modules/python3/processors/pythonscriptsource.cpp]: BasicMesh->Mesh?

File: modules/python3/processors/pythonscriptsource.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ * ### Properties
+ *   * __File Name__  file name of the python script
+ */
+
+/**
+ * \class PythonScriptSource
+ * \brief Loads a mesh and volume via a python script. The processor is invalidated
+ * as soon as the script changes on disk.
+ */
+class IVW_MODULE_PYTHON3_API PythonScriptSource : public Processor {
```
Comment (ResearchDaniel) [inline modules/python3/processors/pythonscriptsource.h]: Name sounds like it should output python script.
Possible to make it template based and separate Mesh and Volume?
Other procedural source processors are using Create, i.e. VolumeCreate. 
VolumePythonCreate?

File: modules/python3/processors/pythonscriptsource.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    "Python",                         // Category
+    CodeState::Experimental,          // Code state
+    "Python, CPU",                    // Tags
+};
+const ProcessorInfo PythonScriptSource::getProcessorInfo() const { return processorInfo_; }
+
+PythonScriptSource::PythonScriptSource()
+    : Processor()
+    , script_("")
+    , scriptFileName_("scriptFileName", "File Name", "")
```
Comment (ResearchDaniel) [inline modules/python3/processors/pythonscriptsource.cpp]: Would be great if it is possible to have a default file that shows how it should be used. Possible future PR

File: modules/python3/pythonscript.h [hunk 4]
```diff
@@ -24,112 +24,127 @@
...
-        std::string filename_;
-        void* byteCode_;
-        bool isCompileNeeded_;
-    };
+    bool run(const VariableMap& extraLocalVariables = VariableMap(),
+             std::function<void(pybind11::dict)> callback = [](pybind11::dict dict) {});
+
+    bool run(std::function<void(pybind11::dict)> callback) { return run({}, callback); }
+
+    virtual void setFilename(const std::string& filename);
```
Comment (ResearchDaniel) [inline modules/python3/pythonscript.h:95]: Ideally, file location should not be needed at script level. Not related to this PR I guess

**Review by ResearchDaniel — DISMISSED** [review]
Nice functionality! 
Have some comments on naming conventions we have used for other procedural source processors

**Comment by martinfalk** [comment]
@petersteneteg suggests `PythonScriptProcessor` as a name candidate

Future work:
- make inports and outports of processor depending on python script
- use secondary script to set up ports
- provide example script

File: modules/python3/processors/pythonscriptsource.h (hunk 2 above)
Comment (martinfalk) [reply modules/python3/processors/pythonscriptsource.h]: no, disagree with separating. No need for templating either. This processor is the first prototype for pulling in data via python, i.e. not creating it per se.

A more advanced version of the processor will determine its ports automatically.

File: modules/python3/processors/pythonscriptsource.cpp (hunk 3 above)
Comment (martinfalk) [reply modules/python3/processors/pythonscriptsource.cpp]: sure, can do.

File: modules/python3/pythonscript.h (hunk 4 above)
Comment (martinfalk) [reply modules/python3/pythonscript.h:95]: that was already there, added by @r-englund earlier.

File: modules/python3/processors/pythonscriptsource.cpp (hunk 1 above)
Comment (martinfalk) [reply modules/python3/processors/pythonscriptsource.cpp]: inviwopy provides both `Mesh` and `BasicMesh`. But only `BasicMesh` provides the functionality for adding vertices like this (at least at the moment until the decorated mesh PR is approved and integrated into inviwopy):
```python
mesh = BasicMesh()
for p in pos :
    mesh.addVertex(vec3(p[0], p[1], p[2]), vec3(0), vec3(0), vec4(1, 0, 0, 1)
```

File: modules/python3/processors/pythonscriptsource.h (hunk 2 above)
Comment (martinfalk) [reply modules/python3/processors/pythonscriptsource.h]: Regarding naming the processor, open for suggestions but they need to be generic.

**Comment by ResearchDaniel** [comment]
Agree on PythonScriptProcessor. Especially if it will be generic.

**Comment by petersteneteg** [comment]
Ideally we would want to be able to save them to the processor list just like the composites

File: modules/python3/pythonscript.h (hunk 4 above)
Comment (martinfalk) [reply modules/python3/pythonscript.h:95]: I am a bit torn here. I can see use cases for both a string/script property as well as an external file.

File: modules/python3/processors/pythonscriptsource.cpp (hunk 3 above)
Comment (martinfalk) [reply modules/python3/processors/pythonscriptsource.cpp]: added example workspace / script for generating a random volume

image

**Review by petersteneteg — APPROVED** [review]
We have a lot more work to do here, but think that can be an other PR...

**Pull request merged**